python/felis/tap_schema.py

```diff
@@ -0,0 +1,633 @@
+        existing database.
+        This can be a mock connection or None, in which case the internal
+        TAP_SCHEMA schema will be used by loading an internal YAML file.
+    schema_name
+        The name of the schema to use for the TAP_SCHEMA tables.
+        Leave as None to use the standard name of "TAP_SCHEMA".
+    apply_schema_to_metadata
+        If True, apply the schema to the metadata as well as the tables.
+        If False, these will be set to None, e.g., for sqlite.
+    table_name_postfix
+        A string to append to all the standard table names.
+        This needs to be used in a way such that the resultant table names
+        map to tables within the TAP_SCHEMA database.
+
+    Notes
+    -----
+    The TAP_SCHEMA schema must either have been created already, in which case
+    the ``engine`` should be provided. Or the internal TAP_SCHEMA schema will
+    be used if ``engine`` is None or a ``MockConnection``.
+    """
+
+    _TABLE_NAMES_STD = ["schemas", "tables", "columns", "keys", "key_columns"]
+    """The standard table names for the TAP_SCHEMA tables."""
+
+    _SCHEMA_NAME_STD = "TAP_SCHEMA"
+    """The standard schema name for the TAP_SCHEMA tables."""
+
+    def __init__(
+        self,
+        engine: Engine | MockConnection | None = None,
+        schema_name: str | None = None,
+        apply_schema_to_metadata: bool = True,
+        table_name_postfix: str = "",
+    ):
+        """Initialize the table manager."""
+        self.table_name_postfix = table_name_postfix
+        self.apply_schema_to_metadata = apply_schema_to_metadata
+        self.schema_name = schema_name or TableManager._SCHEMA_NAME_STD
+
+        if is_valid_engine(engine):
+            assert isinstance(engine, Engine)
+            logger.debug(
+                "Reflecting TAP_SCHEMA database from existing database at %s",
+                engine.url._replace(password="***"),
+            )
+            self._reflect(engine)
+        else:
+            self._load_yaml()
+
+        self._create_table_map()
+        self._check_tables()
+
+    def _reflect(self, engine: Engine) -> None:
+        """Reflect the TAP_SCHEMA database tables into the metadata.
+
+        Parameters
+        ----------
+        engine
+            The SQLAlchemy engine to use to reflect the tables.
+        """
+        self._metadata = MetaData(schema=self.schema_name if self.apply_schema_to_metadata else None)
+        try:
+            self.metadata.reflect(bind=engine)
+        except SQLAlchemyError as e:
+            logger.error("Error reflecting TAP_SCHEMA database: %s", e)
+            raise
+
+    def _load_yaml(self) -> None:
+        """Load the standard TAP_SCHEMA schema from a Felis package resource.
+
+        Parameters
+        ----------
+        yaml_file
+            The path to the YAML file to load.
+        """
+        self._load_schema()
+        if self.schema_name != TableManager._SCHEMA_NAME_STD:
```

Comment:
```suggestion
        if self.schema_name.upper() != TableManager._SCHEMA_NAME_STD:
```

Reply:
Schema names are case sensitive so need to use as is.